blk/components/virt.c

```diff
@@ -184,201 +160,186 @@ static bool handle_mbr_reply()
+        microkit_notify(clients[reqbk.cli_id].ch);
     }
 }
 
 static void handle_client(int cli_id)
 {
     blk_queue_handle_t h = clients[cli_id].queue_h;
-    uintptr_t cli_data_base = blk_virt_cli_data_region(blk_client_data_start, cli_id);
+    uintptr_t cli_data_base = blk_virt_cli_data_region(blk_client_data, cli_id);
     uint64_t cli_data_region_size = blk_virt_cli_data_region_size(cli_id);
 
-    blk_req_code_t cli_code;
-    uintptr_t cli_offset;
-    uint32_t cli_block_number;
-    uint16_t cli_count;
-    uint32_t cli_req_id;
-
-    uintptr_t drv_addr;
-    uint32_t drv_block_number;
-    uint32_t drv_req_id = 0;
+    blk_req_code_t cli_code = 0;
+    uintptr_t cli_offset = 0;
+    uint32_t cli_block_number = 0;
+    uint16_t cli_count = 0;
+    uint32_t cli_req_id = 0;
 
     int err = 0;
     while (!blk_queue_empty_req(&h)) {
         err = blk_dequeue_req(&h, &cli_code, &cli_offset, &cli_block_number, &cli_count, &cli_req_id);
         assert(!err);
 
+        blk_resp_status_t resp_status = BLK_RESP_ERR_UNSPEC;
+
+        if (ialloc_full(&ialloc)) {
+            LOG_BLK_VIRT_ERR("out of request indices to allocate\n");
+            resp_status = BLK_RESP_ERR_UNSPEC;
+            goto req_fail;
+        }
+
+        if (blk_queue_full_req(&drv_h)) {
+            LOG_BLK_VIRT_ERR("driver request queue is full\n");
+            resp_status = BLK_RESP_ERR_UNSPEC;
+            goto req_fail;
+        }
+
+        uint32_t drv_block_number = 0;
         drv_block_number = cli_block_number + (clients[cli_id].start_sector / (BLK_TRANSFER_SIZE / MSDOS_MBR_SECTOR_SIZE));
 
-        if (cli_code == BLK_REQ_READ || cli_code == BLK_REQ_WRITE) {
-            // Check if client request is within its allocated bounds
+        switch (cli_code) {
+        case BLK_REQ_READ:
+        case BLK_REQ_WRITE: {
             unsigned long client_sectors = clients[cli_id].sectors / (BLK_TRANSFER_SIZE / MSDOS_MBR_SECTOR_SIZE);
             unsigned long client_start_sector = clients[cli_id].start_sector / (BLK_TRANSFER_SIZE / MSDOS_MBR_SECTOR_SIZE);
             if (drv_block_number < client_start_sector || drv_block_number + cli_count > client_start_sector + client_sectors) {
+                /* Requested block number out of bounds */
                 LOG_BLK_VIRT_ERR("client %d request for block %d is out of bounds\n", cli_id, cli_block_number);
-                err = blk_enqueue_resp(&h, BLK_RESP_ERR_INVALID_PARAM, 0, cli_req_id);
-                assert(!err);
-                continue;
+                resp_status = BLK_RESP_ERR_INVALID_PARAM;
+                goto req_fail;
             }
 
-            // Check if client request offset is within its allocated bounds and is aligned to transfer size
-            if (cli_offset % BLK_TRANSFER_SIZE != 0 || (cli_offset + BLK_TRANSFER_SIZE * cli_count) > cli_data_region_size) {
+            if ((cli_offset + BLK_TRANSFER_SIZE * cli_count) > cli_data_region_size) {
```

Comment:
This should be `PAGE_SIZE` instead of `BLK_TRANSFER_SIZE`?

Reply:
They are meant to be identical as specified in the sddf design doc, i do think its clearer but i also wanted to avoid adding another #define

Reply:
Oh, I thought `BLK_TRANSFER_SIZE` as decided by the user? Is it always 0x1000?

Reply:
Well its meant to be platform page size

Reply:
Right, but the user would never say like use 0x2000 instead if they wanted larger transfer sizes?

Reply:
yes user is not meant to define it